### PR TITLE
Bluetooth: GATT: Add bt_gatt_write_without_rsp_cb

### LINKED ISSUE #11558: Need notification when a BLE packet is sent
Hi,

We would like to have a notification when a BLE GATT packet is sent.

Our current use-case is to implement accurate clock sync between BLE devices. But I am guessing that there are also many other applications where you want to know if and when a packet is sent.
We have partly solved the problem by calling a hook function from att_pdu_sent() in subsys/bluetooth/host/att.c.
Then we can get a notification when a packet is sent that was previously queued by bt_gatt_write_without_response().

A better solution would be if there was an API where you could add extra parameters to bt_gatt_write_without_response() and also bt_gatt_notify(). And possibly the other sending functions as well.

The extra parameters could be a pointer to a callback function and a void* to some user data.
And then that callback would be called when the transmission of the packet completed.

Or you could register the callbacks once and then only supply the user data when sending the packet.

The problem with the current API is that after requesting to send some data you really have no idea of when sending actually takes place. You only know that it was put in the queue and will be sent some time in the future. But this small change to the API would open up the possibilities to implement more time critical bluetooth applications.

### PR DESCRIPTION
**Pull request by Vudentz**
This adds bt_gatt_write_response_cb works similarly to bt_gatt_notify_cb which can take a callback to be called when the PDU is considered transmitted over the air.

Fixes #11558

### CODE REVIEW

**Comment by Vudentz** [comment]
@m-nilsson Can you take a look?

File: include/bluetooth/gatt.h [hunk 1]
```diff
@@ -995,6 +999,29 @@ struct bt_gatt_write_params {
...
+ * Note: By using a callback it also disable the internal flow control
+ * which would prevent sending multiple commands without wainting their
+ * transmissions to complete, so if that is required the caller shall not
+ * submit more data until the callback is called.
+ *
+ * @param conn Connection object.
+ * @param handle Attribute handle.
+ * @param data Data to be written.
+ * @param length Data length.
+ * @param sign Whether to sign data
```
Comment (jhedberg) [inline include/bluetooth/gatt.h]: The 'func' parameter seems to be undocumented here

File: subsys/bluetooth/host/att.c [hunk 2]
```diff
@@ -2205,11 +2205,14 @@ int bt_att_send(struct bt_conn *conn, struct net_buf *buf)
...
 	}
 
-	k_sem_take(&att->tx_sem, K_FOREVER);
-	if (!att_is_connected(att)) {
-		BT_WARN("Disconnected");
-		k_sem_give(&att->tx_sem);
-		return -ENOTCONN;
+	/* Don't use tx_sem if caller has set it own callback */
+	if (!cb) {
+		k_sem_take(&att->tx_sem, K_FOREVER);
```
Comment (jhedberg) [inline subsys/bluetooth/host/att.c]: If you're not doing `k_sem_take()` when cb is set, then there's somewhere where you also shouldn't be doing `k_sem_give()` if cb was set, right? Where is that place? I don't see it modified in this PR.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/att.c (hunk 2 above)
Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: Yep, the k_sem_give is called only on the stack cbs thus they won't be called if caller has set its own callback.

File: include/bluetooth/gatt.h (hunk 1 above)
Comment (Vudentz) [reply include/bluetooth/gatt.h]: Will fix it.

**Comment by m-nilsson** [comment]
> @m-nilsson Can you take a look?

Looks good!

**Review by jhedberg — APPROVED** [review]

**Pull request merged**